### PR TITLE
Deprecate float sizes (min_size, max_size)

### LINKED ISSUE #1619: Deprecate floating-point values for size arguments
https://github.com/HypothesisWorks/hypothesis/blob/859069c52ccf041234971f30aedaf058d91e7bb7/hypothesis-python/src/hypothesis/internal/validation.py#L102-L111

Sizes - the number of elements in a collection, bounded by `min_size` and `max_size` arguments, should always be integers.  However, we currently also accept floats for these arguments!  We should replace the `check_type(integer_types + (float,), value)` with

```python
if isinstance(value, float):
    note_deprecation(...)
else:
    check_type(integer_types, value) 
```

And then write a test in `hypothesis-python/tests/cover/test_validation.py` that uses `@checks_deprecated_behaviour`, passes `max_size=5.0` to `lists()`, and checks that the resulting lists are at most five elements long.

Finally, we should update the type hints for `average_size` arguments to `# type: None` as `None` is the only non-deprecated value.

### PR DESCRIPTION
**Pull request by math2001**
Will close #1619 

It shows a deprecation warning when `min_size` or `max_size` is a **float**, instead of being an integer.

The tests only target "whole" float number like `5.0`, and not `5.1`.

I tried to find somewhere in the **documentation** where to mention this deprecation, but I didn't find anything... I'm not really sure if it'd be that useful anyway...

It also changes the **type hints** for `average_size` to being `None` instead of `int`.

However, this PR isn't quite finished, I'm not sure about something: **NaN** shouldn't be allowed as a size.

However, a test (in `cover/test_direct_strategies.py`, at `test_validates_keyword_arguments`), checks a bunch of invalid arguments and asserts that they raise `InvalidArgument`.

One of them tests `min_size` as `float('nan')`. And this **breaks** because it raises a warning that it doesn't expect (the warning that this PR adds). Here's the error message

```
Traceback (most recent call last):                                                                                                                                                                
  File "/home/math2001/code/python/hypothesis/hypothesis-python/tests/cover/test_direct_strategies.py", line 158, in test_validates_keyword_arguments                                             
    fn(**kwargs).example()                                                                                                                                                                        
  File "/home/math2001/code/python/hypothesis/hypothesis-python/src/hypothesis/searchstrategy/strategies.py", line 304, in example                                                                
    phases=tuple(set(Phase) - {Phase.shrink}),                                                                                                                                                    
  File "/home/math2001/code/python/hypothesis/hypothesis-python/src/hypothesis/core.py", line 985, in find                                                                                        
    specifier.validate()                                                                                                                                                                          
  File "/home/math2001/code/python/hypothesis/hypothesis-python/src/hypothesis/searchstrategy/strategies.py", line 379, in validate                                                               
    self.do_validate()                                                                                                                                                                            
  File "/home/math2001/code/python/hypothesis/hypothesis-python/src/hypothesis/searchstrategy/lazy.py", line 127, in do_validate                                                                  
    w = self.wrapped_strategy                                                                                                                                                                     
  File "/home/math2001/code/python/hypothesis/hypothesis-python/src/hypothesis/searchstrategy/lazy.py", line 113, in wrapped_strategy                                                             
    *self.__args, **self.__kwargs                                                                                                                                                                 
  File "/home/math2001/code/python/hypothesis/hypothesis-python/src/hypothesis/strategies.py", line 631, in lists                                                                                 
    check_valid_sizes(min_size, average_size, max_size)                                                                                                                                           
  File "/home/math2001/code/python/hypothesis/hypothesis-python/src/hypothesis/internal/validation.py", line 156, in check_valid_sizes                                                            
    check_valid_size(min_size, 'min_size')                                                                                                                                                        
  File "/home/math2001/code/python/hypothesis/hypothesis-python/src/hypothesis/internal/validation.py", line 119, in check_valid_size                                                             
    name, value                                                                                                                                                                                   
  File "/home/math2001/code/python/hypothesis/hypothesis-python/src/hypothesis/_settings.py", line 873, in note_deprecation                                                                       
    warnings.warn(warning, stacklevel=2)                                                                                                                                                          
hypothesis.errors.HypothesisDeprecationWarning: Float size are deprecated: min_size should be an integer, got nan
```

I'm not sure what should be done in this case. Here are the different options I'm thinking of:

1. remove this particular test row (pretty bad)
2. block all floats except NaN so that the exception alone is raised, and not the warning (pretty bad from a user perspective)
3. Find a way for this test to ignore the warning, without breaking all the other tests in the table.

The last solution is, I think, what should be done, but I'm not quite sure how to do it cleanly.

I've tried to use `@pytest.mark.filterwarnings`, without success...

If you want to try it locally:

```
git remote add mathieu https://github.com/math2001/hypothesis
git fetch mathieu deprecate-float-sizes
git checkout deprecate-float-sizes
pytest "hypothesis-python/tests/cover/test_direct_strategies.py::test_validates_keyword_arguments" 
```

### CODE REVIEW

File: hypothesis-python/src/hypothesis/internal/validation.py [hunk 1]
```diff
@@ -106,14 +106,20 @@ def check_valid_size(value, name):
...
+    if isinstance(value, float):
+        note_deprecation(
+            'Float size are deprecated: '
+            '{} should be an integer, got {}'.format(name, value)
+        )
+    else:
+        check_type(integer_types, value)
     if value < 0:
         raise InvalidArgument(u'Invalid size %s=%r < 0' % (name, value))
     if isinstance(value, float) and math.isnan(value):
```
[the remark this answers is not in the document]
Comment (math2001) [reply hypothesis-python/src/hypothesis/internal/validation.py]: Hum... I don't know how I didn't see this... 😢  I'll make sure to read the whole function next time 👀

File: hypothesis-python/src/hypothesis/internal/validation.py [hunk 2]
```diff
@@ -106,14 +106,20 @@ def check_valid_size(value, name):
 
     Otherwise raises InvalidArgument.
     """
+    from hypothesis._settings import note_deprecation
```
[the remark this answers is not in the document]
Comment (math2001) [reply hypothesis-python/src/hypothesis/internal/validation.py]: Oh... I was almost thinking of putting it at the top level 😨. I'll remember this, thanks 🙂

[the remark this answers is not in the document]
Comment (Zac-HD) [reply hypothesis-python/src/hypothesis/internal/validation.py]: Global imports are always better!

*Unless* they cause an import cycle, where in a.py we have `import b` and in b.py, `from a import thing`.  The validation module is used widely enough that basically all our imports have to be inside functions here 😭

File: hypothesis-python/tests/cover/test_direct_strategies.py [hunk 3]
```diff
@@ -154,8 +154,12 @@ def fn_ktest(*fnkwargs):
                      'blacklist_categories': ['Nd']}),
 )
 def test_validates_keyword_arguments(fn, kwargs):
+    from hypothesis.errors import HypothesisDeprecationWarning
     with pytest.raises(InvalidArgument):
-        fn(**kwargs).example()
+        # pytest.warns filters out the warning. It doesn't ensure that it
```
[the remark this answers is not in the document]
Comment (math2001) [reply hypothesis-python/tests/cover/test_direct_strategies.py]: This did seem weird... 🤔 And, I messed around with it, and I think there is a bit of an edge case here, when `pytest.warns` and `pytest.raises` are nested...

```python
import warnings
import pytest

def should_warn(): # but doesn't
    raise KeyError('...')

with pytest.raises(KeyError):
    with pytest.warns(UserWarning):
        warns()
```

This doesn't raise any error (python `3.7.0`, pytest `3.8.2`)

But when I tested:

```python
def should_warn():
    pass

with pytest.warns(UserWarning):
    warn()
```

This does raise an error.

What am I missing here?

> You shouldn't need to change this test at all though! If any entries use floats for a size argument, delete them here - you've already written the new tests with @checks_deprecated_behaviour instead.

You mean that I should change the test data? Wouldn't that break the compatibility? People would expect to get an `InvalidArgument`, yet we wouldn't be *testing* it (although it would still work). 

Anyway, this is resolved with your comment below 😄

File: hypothesis-python/src/hypothesis/internal/validation.py [hunk 4]
```diff
@@ -113,11 +113,18 @@ def check_valid_size(value, name):
                 'min_size=None is deprecated; use min_size=0 instead.'
             )
         return
-    check_type(integer_types + (float,), value)
-    if value < 0:
-        raise InvalidArgument(u'Invalid size %s=%r < 0' % (name, value))
     if isinstance(value, float) and math.isnan(value):
```
Comment (Zac-HD) [inline hypothesis-python/src/hypothesis/internal/validation.py]: It's more efficient in the non-deprecated case to write

```python
if isinstance(value, float):
    if math.isnan(value):
        raise ...
    from ... import ...
```

so we only ever check whether the value is a float once.

**Comment by math2001** [comment]
Whooo! Thank you!

I'll have a look at #1600.
